### PR TITLE
[SPARK-31077][ML] Remove ChiSqSelector dependency on mllib.ChiSqSelectorModel

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?

```ChiSqSelector ``` depends on ```mllib.ChiSqSelectorModel``` to do the selection logic. Will remove the dependency in this PR. 

### Why are the changes needed?
This PR is an intermediate PR.  Removing ```ChiSqSelector``` dependency on ```mllib.ChiSqSelectorModel```. Next subtask will extract the common code between ```ChiSqSelector``` and ```FValueSelector``` and put in an abstract ```Selector```. 

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
New and existing tests

### CODE REVIEW

**Comment by srowen** [comment]
General questions: does this mean not being able to share as much code? or how do we avoid parallel implementations if one doesn't depend on the other -- they can both share common code elsewhere?
Also why do some of the binary test files change?

**Comment by huaxingao** [comment]
@srowen Thank you very much for your review!

I am trying to remove the dependency on mllib.ChiSqSelectorModel. After the change, ChiSqSelector.fit, ChiSqSelectorModel  constructor and implementation are almost the same as the FValueSelector implementation, then I will extract all the common code between ```ChiSqSelector``` and ```FValueSelector``` to put in an abstract ```Selector```. 

So basically here is the final goal: 

```
abstract class Selector[T <: SelectorModel[T]]
it contains all the params, fit method + 2 abstract methods:

// Compute the statistics information (pValue + chisq statistics/fvalue)
protected[this] def getSelectionTestResult(dataset: Dataset[_]): Array[SelectionTestResult]

// Create SelectorModel (could be ChiSqModel or FValueModel or ANOVAModel)
protected[this] def createSelectorModel(
      uid: String,
      statistics: Array[Double]): T
```

Then I will have three concrete Selector classes:
```
1. 
// categorical features and categorical labels.
class ChiSqSelector extends Selector[ChiSqSelectorModel]
implement getSelectionTestResult to return pValue and chisq stats
implement createSelectorModel to create a ChiSqSelectorModel

2.
// continuous features and continuous labels
class FValueSelector extends Selector[FValueSelectorModel]
implement getSelectionTestResult to return pValue and f regression value
implement FValueSelector to create a FValueSelectorModel

3.
// continuous features and categorical labels
class ANOVASelector extends Selector[ANOVASelectorModel]
implement getSelectionTestResult to return pValue and ANOVA fvalue
implement ANOVASelector to create a ANOVASelectorModel
```

The initial PR is https://github.com/apache/spark/pull/27527. I broke it to several small PRs so it will be easier to review:

1. add stat.FValueRegressionTest (https://github.com/apache/spark/pull/27623)
2. add feature.FValueSelector (https://github.com/apache/spark/pull/27679)
3. this PR to remove ChiSqSelector dependency on mllib.ChiSqSelectorModel. 
4. extract all the common code to abstract Selector. Make FValueSelector and ChiSqSelector implement abstract Selector.
5. Add ANOVASelector (the implementation of ANOVA is not in https://github.com/apache/spark/pull/27527)


Maybe I should combine step 3 and 4 in one PR?

For the binary test files changes: I initially planned to exposed pValue and chisq statistics in ChisqSelectorModel so the constructor is like this:
```
final class ChiSqSelectorModel private[ml] (
    @Since("1.6.0") override val uid: String,
    override val selectedFeatures: Array[Int],
    @Since("3.1.0") override val pValues: Array[Double],
    @Since("3.1.0")override val statistic: Array[Double])
```
Then when I load the model, I will need to do this:
```
      val model = if (majorVersion < 3 || (majorVersion == 3 && minorVersion < 1)) {
        // model prior to 3.1.0
        val data = df.select("selectedFeatures").head()
        ...
      } else {
        val data = df.select("selectedFeatures", "pValue", "statistics").head()
       ...
      }
```
So I have the binary test files to test the prior 3.1.0 model. But then I wasn't so sure if I need to expose pValue and chisq statistics, and I decided to rethink this later in the final step when I do the abstract Selector. I changed the constructor and reader code but forgot to remove the binary test files. I will clean up now. Sorry my mistake confused you.

**Comment by srowen** [comment]
Up to you how you want to proceed. 
OK so in the end this should increase code sharing; it's not separating the implementations by copying.
p values do sound useful in the model; up to you. Yes it just needs to be backwards-compatible.

**Comment by huaxingao** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 1]
```diff
@@ -245,14 +266,16 @@ object ChiSqSelector extends DefaultParamsReadable[ChiSqSelector] {
...
-    private val chiSqSelector: feature.ChiSqSelectorModel)
+    @Since("3.1.0")val selectedFeatures: Array[Int])
   extends Model[ChiSqSelectorModel] with ChiSqSelectorParams with MLWritable {
 
   import ChiSqSelectorModel._
 
-  /** list of indices to select (filter). */
-  @Since("1.6.0")
-  val selectedFeatures: Array[Int] = chiSqSelector.selectedFeatures
+  var prev = -1
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: Doesn't matter, but you can write `require(selectedFeatures.sliding(2).forall(l => l(0) < l(1)), ...)`

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 2]
```diff
@@ -270,10 +293,10 @@ final class ChiSqSelectorModel private[ml] (
     val func = { vector: Vector =>
       vector match {
         case SparseVector(_, indices, values) =>
-          val (newIndices, newValues) = chiSqSelector.compressSparse(indices, values)
+          val (newIndices, newValues) = compressSparse(indices, values)
           Vectors.sparse(newSize, newIndices, newValues)
         case DenseVector(values) =>
-          Vectors.dense(chiSqSelector.compressDense(values))
+          Vectors.dense(selectedFeatures.map(i => values(i)))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: does it work to just `.map(values)`?

File: mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala [hunk 3]
```diff
@@ -75,4 +77,34 @@ object ChiSquareTest {
...
+  /**
+   * @param dataset  DataFrame of categorical labels and categorical features.
+   *                 Real-valued features will be treated as categorical for each distinct value.
+   * @param featuresCol  Name of features column in dataset, of type `Vector` (`VectorUDT`)
+   * @param labelCol  Name of label column in dataset, of any numerical type
+   * @return Array containing the SelectionTestResult for every feature against the label.
+   */
+  @Since("3.1.0")
+  def testChiSquare(dataset: Dataset[_], featuresCol: String, labelCol: String):
+  Array[SelectionTestResult] = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala]: Nit: break this a little differently? indent more here or wrap the arguments and indent

File: mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala [hunk 4]
```diff
@@ -75,4 +77,34 @@ object ChiSquareTest {
...
+    SchemaUtils.checkColumnType(dataset.schema, featuresCol, new VectorUDT)
+    SchemaUtils.checkNumericType(dataset.schema, labelCol)
+    val input: RDD[OldLabeledPoint] =
+      dataset.select(col(labelCol).cast(DoubleType), col(featuresCol)).rdd
+        .map {
+          case Row(label: Double, features: Vector) =>
+            OldLabeledPoint(label, OldVectors.fromML(features))
+        }
+    val chiTestResult = OldStatistics.chiSqTest(input)
+    var chiTestResultArray = new Array[SelectionTestResult](chiTestResult.length)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala]: Another nit but I think the remainder of the function could be:
```
chiTestResult.map(r => new ChiSqTestResult(r.pValue, r.degreesOfFreedom, r.statistic)
```

**Review by srowen — COMMENTED** [review]
Looks OK; ready to merge otherwise?
just a few non-essential suggestions

**Comment by huaxingao** [comment]
Fixed. Thank you very much! @srowen

File: mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala [hunk 5]
```diff
@@ -75,4 +77,31 @@ object ChiSquareTest {
...
   }
+
+  /**
+   * @param dataset  DataFrame of categorical labels and categorical features.
+   *                 Real-valued features will be treated as categorical for each distinct value.
+   * @param featuresCol  Name of features column in dataset, of type `Vector` (`VectorUDT`)
+   * @param labelCol  Name of label column in dataset, of any numerical type
+   * @return Array containing the SelectionTestResult for every feature against the label.
+   */
+  @Since("3.1.0")
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala]: That difference between this method and above `def test(dataset: DataFrame, featuresCol: String, labelCol: String): DataFrame` is only that the result are parallelized to a dataframe?

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 6]
```diff
@@ -245,14 +266,15 @@ object ChiSqSelector extends DefaultParamsReadable[ChiSqSelector] {
 @Since("1.6.0")
 final class ChiSqSelectorModel private[ml] (
     @Since("1.6.0") override val uid: String,
-    private val chiSqSelector: feature.ChiSqSelectorModel)
+    @Since("3.1.0")val selectedFeatures: Array[Int])
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: nit: missing space before `val`

File: mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala [hunk 7]
```diff
@@ -75,4 +77,31 @@ object ChiSquareTest {
...
+  def testChiSquare(
+      dataset: Dataset[_],
+      featuresCol: String,
+      labelCol: String): Array[SelectionTestResult] = {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala]: What about returning `Array[ChiSquareResult]`?

**Review by zhengruifeng — COMMENTED** [review]
I am OK with it as a migration of `ChisqSelector`  to the `.ml` side .

File: mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala (hunk 5 above)
Comment (huaxingao) [reply mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala]: Yes. 
I may need to refactor this method and move it out of this class in next subtask.

File: mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala (hunk 7 above)
Comment (huaxingao) [reply mllib/src/main/scala/org/apache/spark/ml/stat/ChiSquareTest.scala]: I will need to have an abstract method in abstract class ```Selector``` in next subtask
```
def getSelectionTestResult(dataset: Dataset[_]): Array[SelectionTestResult]
```
so I prefer to return ```Array[SelectionTestResult]``` here.

**Comment by huaxingao** [comment]
retest this please

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by srowen** [comment]
If you're OK with this in its current form I will merge.

**Comment by huaxingao** [comment]
Please merge. Thank you very much! @srowen

**Comment by srowen** [comment]
Merged to master

**Comment by huaxingao** [comment]
Thanks! @srowen @zhengruifeng

**Pull request closed without merging**